An extra sandbox container created after job pod succeed

liucimin opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

 /kind bug


**What this PR does / why we need it**:

When a job pod finish quickly.
Such as the container's args is "date; echo Hello from the Kubernetes cluster"
the container and sandbox will exited quickly,
then sometimes if the RECONCILE event slower than the PLEG ContainerDied event kubelet will start a new sandbox.
It because the kubelet worker use the old pod status from the podmanager 
So it will create two sandbox when a job finish.
The first sandbox will teardown when in the PLEG ContainerDied Even syncpod.
The second sandbox will teardown by housekeeping.
But the second sandbox dont have network namespace.So the cni can not release some network 
resource.
So we should try to not create two sandbox in one job,

Check if the pod's container all succeeded,the kubelet will not create a new sandbox.


**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #75441 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
NONE
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->

```release-note
When check sandbox changed,filter the succeeded pod.
```

liucimin commented:
/sig node
/area kubelet

dcbw commented:
/ok-to-test

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -417,25 +417,39 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *v1.Pod, podStatus *ku
+	//check the pod is succeeded
+	succeededContainerCount := 0
+	for _, container := range pod.Spec.Containers {
+		for _, status := range podStatus.ContainerStatuses {
```
Why not just iterate over the ContainerStatuses?

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -417,25 +417,39 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *v1.Pod, podStatus *ku
 	}
-	if sandboxStatus.State != runtimeapi.PodSandboxState_SANDBOX_READY {
+
+	//Needs to create a new when not all container succeeded and the ready sandbox is not the latest one
```
Space after //
create a new what?
```suggestion
	// Needs to create a new ??? when not all container succeeded and the ready sandbox is not the latest one
```

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -417,25 +417,39 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *v1.Pod, podStatus *ku
 
-	// Needs to create a new sandbox when network namespace changed.
-	if sandboxStatus.GetLinux().GetNamespaces().GetOptions().GetNetwork() != networkNamespaceForPod(pod) {
+	// Needs to create a new sandbox when not all container succeeded and network namespace changed.
```
nit: I would rephrase this as
```suggestion
	// Needs to create a new sandbox when network namespace changed, unless all containers are succeeded.
```

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -417,25 +417,39 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *v1.Pod, podStatus *ku
-	if sandboxStatus.State != runtimeapi.PodSandboxState_SANDBOX_READY {
+
+	//Needs to create a new when not all container succeeded and the ready sandbox is not the latest one
+	if len(pod.Spec.Containers) != succeededContainerCount && sandboxStatus.State != runtimeapi.PodSandboxState_SANDBOX_READY {
```
Just return early if `len(pod.Spec.Containers) == succeededContainerCount`

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -417,25 +417,39 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *v1.Pod, podStatus *ku
+				}
+			}
+		}
+	}
```
Move this block after the readySandboxCount check

liucimin replied:
Ok,I have move this block after the readySandboxCount check.

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -417,25 +417,39 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *v1.Pod, podStatus *ku
+	for _, container := range pod.Spec.Containers {
+		for _, status := range podStatus.ContainerStatuses {
+			if container.Name == status.Name {
+				if status.State == kubecontainer.ContainerStateExited && status.ExitCode == 0 {
```
I think this depends on the restart policy? If the restart policy is `Always`, the sandbox should still be recreated. If the restart policy is `Never`, then failed containers should be included.

liucimin replied:
Yes.
Now the code show us that:
when the restart policy is not  Always and the all container is succeded, the sandbox will not create a new one.
Others will enter the old process.

tallclair commented:
/assign

liucimin committed:
when a cronjob or a job finish too quick,
        the container and sandbox will exited quickly,
        then sometimes if the RECONCILE event slower than the PLEG ContainerDied event kubelet will start a new sandbox.
        So it will create two sandbox when a job finish.

liucimin commented:
friendly ping @tallclair

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -417,12 +417,27 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *v1.Pod, podStatus *ku
 		return true, sandboxStatus.Metadata.Attempt + 1, sandboxStatus.Id
 	}
+
+	//check the pod is succeeded
```
nit: Space after `//` (everywhere)

liucimin replied:
Thx.

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -417,12 +417,27 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *v1.Pod, podStatus *ku
+	//check the pod is succeeded
+	succeededContainerCount := 0
+	for _, container := range pod.Spec.Containers {
+		if containerSucceeded(&container, podStatus) {
```
The logic for this method looks wrong to me... I think the container should only be considered succeeded if it's state is `ContainerStateExited`.

/cc @yujuhong

liucimin replied:
Do you mean that there is a problem with the logic of the function of `containerSucceeded`?

tallclair replied:
Yes. In the context we used it before, it didn't matter, but it would in this context.
